README.md

# Defiance #

### Overview ###
This is an enhanced version of the Sydney Engine multiplayer shooter game, written in Java originally by Keith Woodward in 2008, using the Apache Mina networking framework (version 2.0.9).

* Version 1.0.1
* Website: [http://greenlantern101.github.io/Defiance/](http://greenlantern101.github.io/Defiance/)
* Wiki: [https://github.com/GreenLantern101/Defiance/wiki](https://github.com/GreenLantern101/Defiance/wiki)

### Screenshots ###
![After which I succumbed to great firepower.](https://github.com/GreenLantern101/Defiance/blob/master/SydneyShooter/Defiance2.PNG "After which I succumbed to great firepower.")

### Playing ###

======[ Dependencies ]====== 
The jars are included in the SydneyDependencyJars folder.

======[ Running ]====== 
Just load the Eclipse Project, and run the GameFrame.java class.
Alternatively, just run Defiance.jar.


======[ Controls ]====== 

Move with Arrow keys or W, A, S, D
Left mouse click: fires weapon
R: reloads weapon
Number keys (1-9) or Q/E or Mouse wheel: selects available weapons
`[funny key above TAB]: turn on/off player names and hit points
Enter or ShiftEnter: Activate chat, then and once you've typed the message, press enter again to send. 
If shift is down when you first press enter, the 'send to allies' option is ticked
Page up or down: zoom in/ out
Escape: displays menu.
				
======[ Weapons ]====== 

Pistol - Basic weapon with low damage.

Machine Gun - Standard rapid-firing rotating gatling.

Flamethrower - Close-range weapon that sprays napalm that will stick around for a while, useful for area denial. The napalm will cause damage to anyone that walks over it.

Shotgun - Dual-shot weapon that fires a cloud of pellets. Most effective at close ranges.

Sniping Rifle - Accurate for long-distance one-hit kills, but slow-firing.

Tranquilizer Gun - Fires in short bursts, stunning enemies and slowing them and making them unable to fire.

Rocket Launcher - launches rockets that have a large explosion radius. More damage is dealt to objects that are closer to the epicenter of the explosion.

Homing Gun - launches three alien-technology orbs that will fire like a regular bullet unless there is an enemy in front of it, in which case it will automatically follow that target. The orbs are relatively slow, but are still useful around corners and against dug-in defenders.

Nail Gun - An alternative to the machine gun, that spits out 3 nails at once very rapidly, but watch out for your ammo depletion, since it comes with a small clip considering its rate of fire.

***Note***

Both the Flamethrower and the Rocket Launcher are Variable-Range weapons, meaning their bullets can detonate or stick at a  certain range away from the weapon by using the mouse to aim.
				
======[ Items ]====== 

Healthpack - Restores 33% health.

Godly Armor - Grants invincibility against all weapons for 15 seconds, but player can still be tranquilized.

Speed Shoes - Significantly speeds up travel.

Invisibility Shroud - Enemies can't see and target you for 15 seconds.


***Note***

You can only use one of either the Godly Armor or the Invisibility Shroud, but not both. But, the Speedy Shoes can be used with anything.
				
======[ Extra ]======  

The wi-fi on laptops may cause noticeable lag over local LAN, but still very playable. 
				
This was slightly modified from the excellent Sydney Engine v. 0.2, made by Keith Woodward,  whom can find on the forums at www.javagaming.org or contact through keithphw@hotmail.com .

* For more information, please consult the wiki.

### Contribution guidelines ###

* For changes, please consult the ChangeLog.

TODO:

Fix internet gameplay, make networking more dependable

Achievements??? - 1st kill, 3rd, 10th, 20th, 30th, 50th, 75th, 100th, 150th, 300th, 500th kills
Die 1 time,
Die 3 times with no kills, 
Get 3 healthpacks, 
Get 3 invisibilities, 
Switch weapons 20 times, 
used each weapon at least once, 
killed with each weapon at least once, 
assist 25 times

Make personal pop-up alerts configurable in new "GAMEPLAY" menu option

Enhance graphics

Mech Suits support?

Destructable walls- possibly by using subtractive polygons

Map Editor/Maker!

======= when adding a new Gun:

1. Make new gun, and corresponding bullet class

2. Update the statement in Player class of what killed description should be

3. Update all the itemSpawners in gameworld class to spawn it

4. Update bots to spawn it when respawning

5. Update help menu info

### Contacts ###

You can reach me at tiger201655@yahoo.com


ChangeLog.txt
ChangeLog

12/20/2014
============== Gameplay
Enhanced rocket, sniping rifle (faster, larger, damage)
Flamethrower now ignite area of ground on fire for some time
============== UI/ HUD
Added Q/E switch weapon key support
Changed to regular up/down/left/right movement system, and bots override it themselves
public chat messages for kills and killStreak(restarts when you die) and means of death
============== Graphics
Enhanced regular canvas graphics


12/21/2014
============== UI/ HUD
Fully implemented "assists" in kills
Player names now highlighted in chat system
Enhanced healthbar
Small reformatting of Performance Options menu, including show stats option


12/22/2014
============== Gameplay
Flamethrower napalm sticks to walls
Rockets explode into shrapnel
Powerup added: GodlyArmor, which grants invincibility for 15 seconds (can't use more than one powerup at once)
Tranquilizer stop people from firing for a short time, rebalanced Tranquilizer gun
============== UI/ HUD
Limits on zoom in/ zoom out

12/23/2014
=============== Bugfixes
Tranquilized bots now can resume firing later
Bots are now able to use godly armor and invisibility
Bots have secondary armament (pistol), better rocket launcher AI
Bots have basic chat
=============== Gameplay
SpeedShoes item added
implemented personal messages/alerts
Removed item labels, so must run into them to find out what they are!
=============== UI/HUD
Updated Help Menu

12/24/2014
=============== Bugfixes
Networking now works on local LAN, but slight lag for all client players on laptops, but not on desktops
Update code to use Apache Mina 2.0.9 instead of 1.1.5;
Organized imports, removing unnecessary references
=============== Gameplay
Added cheats:
give me speed
give me allguns
give me cloak
give me stealth
=============== Graphics/UI
Bullet trail on pistol, machinegun, rocket

12/25/2014
=============== Gameplay
Players can slide along walls & banks of water now
Can now aim the range of flamethrower and rocket launcher based on mouse
Added homing missile gun
=============== UI/HUD
implemented translucent Arc heads-up-display of ammo in clip
=============== Misc.
better console output for debugging, tracing

12/26/2014
=============== Misc.
Reorganized files
=============== Gameplay
Added Nailgun
Made a Custom Color class for exotic colors like Khaki, Maroon, Turquoise, etc.

12/29/2014
=============== Bugfixes
Fixed threading issue with client join attempt failing immediately.
Server creators allowed to choose IP address they wish to use if they have more than one.
Point2D.Double target issue in Homing Bullet fixed


TODO:
Fix internet gameplay

Achievements??? - 1st kill, 3rd, 10th, 20th, 30th, 50th, 75th, 100th, 150th, 300th, 500th kills
Die 1 time, Die 3 times with no kills, Get 3 healthpacks, Get 3 invisibilities, Switch weapons 20 times, used each weapon at least once, killed with each weapon at least once, assist 25 times

Make alerts configurable

Enhance graphics

Mech Suits support?
Destructable walls- possibly by using subtractive polygons
Map Editor/Maker!

Make bots smarter - auto pathfinding, bullet dodging, finding cover, etc.

======= when adding a new Gun:
1. Make new gun, and bullet class
2. Update the statement in Player class of what killed description should be
3. Update all the itemSpawners in gameworld class to spawn it
4. Update bots to spawn it when respawning
5. Add help menu info



SydneyShooter/src/sydneyengine/shooter/CapturableFlag.java
package sydneyengine.shooter;

import java.awt.BasicStroke;
import java.awt.FontMetrics;
import java.awt.Graphics2D;
import java.awt.Stroke;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Point2D;

import sydneyengine.superserializable.SSAdapter;

public class CapturableFlag extends SSAdapter {

	GameWorld world;
	Team team = null;
	float radius = 20;
	//transient Ellipse2D.Float flagAreaCircle = new Ellipse2D.Float(0, 0, radius * 2f, radius * 2f);
	Player enemyThief = null;	// if not null, this enemy player has the flag.
	boolean dropped = false;
	float dropPointX;
	float dropPointY;
	double dropTime = 0;
	static float dropTimeBeforeReplace = 10;	// seconds
	float homeX;
	float homeY;

	public CapturableFlag() {
	}

	public CapturableFlag(GameWorld world, Team team, float homeX, float homeY) {
		this.world = world;
		this.team = team;
		this.homeX = homeX;
		this.homeY = homeY;
		//flagAreaCircle = new Ellipse2D.Float((getCurrentX() - radius), (getCurrentY() - radius), 2f * radius, 2f * radius);
	}
	
	public void drop(double dropTime){
		if (enemyThief != null) {
			dropped = true;
			enemyThief.setCapturableFlag(null);
			dropPointX = enemyThief.getX();
			dropPointY = enemyThief.getY();
			this.dropTime = dropTime;
			enemyThief = null;
			//System.out.println("CapturableFlag: flag dropped by: " + enemyThief.toString());
		}
	}
	public void doMove(double seconds, double timeAtStartOfMoveSeconds) {
		if (enemyThief != null) {
			if (enemyThief.getTeam().getCapturableFlag().isAtBase() == true){
				//SpawnFlag enemyBaseFlag = enemyThief.getTeam().getSpawnFlagBase();
				for (SpawnFlag enemyBaseFlag : enemyThief.getTeam().getSpawnFlags()){
					float distance = (float) Point2D.distance(enemyBaseFlag.getX(), enemyBaseFlag.getY(), enemyThief.getX(), enemyThief.getY());
					if (distance < enemyThief.getR() + enemyBaseFlag.radius) {
						enemyThief.addFlagCapture();
						//System.out.println("CapturableFlag: flag successfully captured by: " + enemyThief.toString()+".... "+enemyBaseFlag.getX()+", "+enemyBaseFlag.getY()+", "+enemyThief.getX()+", "+enemyThief.getY());
						enemyThief.setCapturableFlag(null);
						enemyThief = null;
						break;
					}
				}
			}
			
		} else if (dropped == true) {
			if (seconds + timeAtStartOfMoveSeconds > dropTime + dropTimeBeforeReplace) {
				dropped = false;
				//System.out.println("CapturableFlag: flag returned after being dropped");
			} else {
				for (Player player : world.getPlayers()) {
					if (player.isDead() == true) {
						continue;
					}
					float distance = (float) Point2D.distance(dropPointX, dropPointY, player.getX(), player.getY());
					if (distance < player.getR() + radius) {
						if (player.getTeam() == getTeam()) {
							player.addFlagReturn();
							dropped = false;
							//System.out.println("CapturableFlag: flag secured from after dropped");
						} else {
							enemyThief = player;
							enemyThief.setCapturableFlag(this);
							dropped = false;
							//System.out.println("CapturableFlag: kidnapped by: " + player + ", after being dropped!");
						}
					}
				}
			}
		} else {
			for (Player player : world.getPlayers()) {
				if (player.isDead() == true || player.getTeam() == getTeam()) {
					continue;
				}
				float distance = (float) Point2D.distance(homeX, homeY, player.getX(), player.getY());
				if (distance < player.getR() + radius) {
					enemyThief = player;
					player.setCapturableFlag(this);
					/*String m1 = "James_Bond.mid";
					String m2 = "Jaws.mid";
					BigFrame.playMusic((Math.random() > 0.5 ? m1 : m2));*/
					//System.out.println("CapturableFlag: kidnapped by: " + player);
				}
			}
		}
		assert dropped == true || enemyThief != null || getCurrentX() == homeX : dropped+", "+enemyThief+", "+getCurrentX()+", "+homeX;
	}

	public void render(ViewPane viewPane) {
		Graphics2D g = viewPane.getBackImageGraphics2D();
		g.setColor(team.getColor());
		/*
		if (dropped){
			g.setColor(Color.MAGENTA);
		}
		else if (enemyThief == null){
			g.setColor(Color.GREEN);
		}else{
			enemyThief.render(viewPane);
			g.setColor(Color.YELLOW);
			g.drawString(enemyThief.getSSCode()+", "+getWorld().getPlayers().toString(), getCurrentX(), getCurrentY()+20);
		}*/
		Stroke oldStroke = g.getStroke();
		g.setStroke(new BasicStroke(4));
		//flagAreaCircle.x = getCurrentX() - radius;
		//flagAreaCircle.y = getCurrentY() - radius;
		g.draw(new Ellipse2D.Float((getCurrentX() - radius), (getCurrentY() - radius), 2f * radius, 2f * radius));
		//g2D.draw(new Ellipse2D.Float((getCentre().x-radius),(getCentre().y-radius),2*radius,2*radius));
		g.setStroke(oldStroke);
		//System.out.println(this.getClass().getSimpleName()+": enemyThief == "+enemyThief+", dropped == "+dropped+", getCurrentX() == "+getCurrentX()+", getCurrentY() == "+getCurrentY()+", getTeam().getSpawnFlagBase().getX() == "+getTeam().getSpawnFlagBase().getX()+", getTeam().getSpawnFlagBase().getY() == "+getTeam().getSpawnFlagBase().getY());
		if (viewPane.isShowMapDescriptions() || containsPoint(viewPane.getWorldMouseX(), viewPane.getWorldMouseY())) {
			// draw the name
			FontMetrics fm = viewPane.getFontMetrics(g.getFont());
			String text = (this.getTeam() == viewPane.getPlayer().getTeam() ? "Your team's flag" : "Enemy team's flag");
			float stringWidth = fm.stringWidth(text) / 2f;
			g.drawString(text, getCurrentX() - stringWidth, getCurrentY()+radius+15);
		}
	}
	public boolean containsPoint(float px, float py) {
		if (Point2D.distance(getCurrentX(), getCurrentY(), px, py) < radius) {
			return true;
		}
		return false;
	}
	public boolean isAtBase(){
		if (enemyThief != null) {
			return false;
		} else if (dropped == true) {
			return false;
		}
		return true;
	}
	public float getCurrentX() {
		float xNow = homeX;
		if (enemyThief != null) {
			xNow = enemyThief.getX();
		} else if (dropped == true) {
			xNow = dropPointX;
		}
		return xNow;
	}

	public float getCurrentY() {
		float yNow = homeY;
		if (enemyThief != null) {
			yNow = enemyThief.getY();
		} else if (dropped == true) {
			yNow = dropPointY;
		}
		return yNow;
	}
	
	public GameWorld getWorld() {
		return world;
	}

	public Team getTeam() {
		return team;
	}
}


SydneyShooter/src/sydneyengine/shooter/Gun/FlameThrower.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package sydneyengine.shooter.Gun;

import java.awt.BasicStroke;
import java.awt.Graphics2D;
import java.awt.Stroke;
import java.awt.geom.Line2D;

import sydneyengine.shooter.CustomColor;
import sydneyengine.shooter.GameWorld;
import sydneyengine.shooter.Player;
import sydneyengine.shooter.ViewPane;
import sydneyengine.shooter.Bullet.Bullet;
import sydneyengine.shooter.Bullet.FlameBall;
import sydneyengine.shooter.Item.Item;
/**
 *
 * @author CommanderKeith
 */
public class FlameThrower extends Gun{
	
	static int playerGunNum = 2;
	float coneAngle;
	
	public FlameThrower(){
		respawn();
	}
	public FlameThrower(GameWorld world) {
		this();
		this.world = world;
	//System.err.println("getSSCode() == "+getSSCode()+", dead == "+dead);
	}
	
	@Override
	public void respawn() {
		super.respawn();
		gunLength = 5;
		reloadSeconds = 0.020f;
		reloadClipSeconds = 2;
		ammo = 0;
		maxAmmoInClip = 500;
		ammoInCurrentClip = maxAmmoInClip;
		maxGunRotationSpeed = (float)Math.PI*2f;
		coneAngle = (float)Math.PI/16f;
	}
	@Override
	public float getRangeForBotAiming(){
		return 300;
	}
	@Override
	public int getPlayerGunNum(){
		return playerGunNum;
	}
	@Override
	public Item createNewItem(GameWorld world){
		return new FlameThrower(world);
	}
	@Override
	public Bullet createBullet(Player player, float xPosWhenFired, float yPosWhenFired, float gunAngle, float mouseX, float mouseY, double lastTimeFiredSeconds, float xLaunchSpeed, float yLaunchSpeed){
		world.getRandom().setSeed(seed);
		seed += 3491;
		//float randomAngleIncrement = (float)((world.getRandom().nextGaussian()-0.5)*coneAngle);
		float randomAngleIncrement = (float)((world.getRandom().nextFloat()-0.5)*coneAngle);
		world.getRandom().setSeed(seed);
		seed += 1237;
		int colorNum = world.getRandom().nextInt(FlameBall.colors.size());
		assert colorNum < FlameBall.colors.size() : colorNum;
		
		return new FlameBall(this, player, xPosWhenFired, yPosWhenFired, gunAngle + randomAngleIncrement, mouseX, mouseY, lastTimeFiredSeconds, xLaunchSpeed, yLaunchSpeed, colorNum);
	}
	
	transient BasicStroke stroke = new BasicStroke(5);
	@Override
	public void render(ViewPane viewPane) {
		super.render(viewPane);
		Graphics2D g = viewPane.getBackImageGraphics2D();
		g.setColor(CustomColor.DARK_ORANGE);
		Stroke oldStroke = g.getStroke();
		g.setStroke(stroke);
		g.draw(new Line2D.Float(player.getX(), player.getY(), (float) (player.getX() + gunLength * Math.cos(gunAngleRelativeToPlayer+player.getAngle())), (float) (player.getY() + gunLength * Math.sin(gunAngleRelativeToPlayer+player.getAngle()))));
		g.setStroke(oldStroke);
	}
	
	
}

SydneyShooter/src/sydneyengine/shooter/Bullet/HomingBullet.java
package sydneyengine.shooter.Bullet;

import java.awt.Graphics2D;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Line2D;
import java.awt.geom.Point2D;
import java.util.ArrayList;

import sydneyengine.shooter.CustomColor;
import sydneyengine.shooter.KPolygon;
import sydneyengine.shooter.Obstacle;
import sydneyengine.shooter.Player;
import sydneyengine.shooter.ViewPane;
import sydneyengine.shooter.Gun.HomingGun;

public class HomingBullet extends Bullet {

	static float canNotHitOwnPlayerTimeSeconds = 1.5f;
	float radius;
	float length;
	float maxSpeed;
	float damage;
	float angle;
	float maxRotationSpeed = (float) Math.PI;
	float startX; float startY;

	protected float lifeTimeSeconds;
	// fixed Point2D.Double problem
	double targetX;
	double targetY;

	public HomingBullet() {
		super();
		player = null;
	}

	public HomingBullet(HomingGun gun, Player player, float newX, float newY, float angle, double spawnTimeSeconds, float xLaunchSpeed, float yLaunchSpeed) {
		super(player, newX, newY, angle, spawnTimeSeconds, xLaunchSpeed, yLaunchSpeed);
		assert Point2D.distance(player.getX(), player.getY(), newX, newY) < player.getR() : Point2D.distance(player.getX(), player.getY(), newX, newY);
		radius = 3.0f;
		length = 2*radius;
		damage = 25.0f;
		this.angle = angle;
		gun.getWorld().getRandom().setSeed(gun.getSeed());
		gun.setSeed(gun.getSeed()+3041);
		float randomSpeedIncrement = world.getRandom().nextFloat()*30;
		float startSpeed = 60 + randomSpeedIncrement;
		gun.getWorld().getRandom().setSeed(gun.getSeed());
		gun.setSeed(gun.getSeed()+1761);
		float randomRangeIncrement = world.getRandom().nextFloat()*200;
		float range = 600 + randomRangeIncrement;
		speedX = xLaunchSpeed + (float) Math.cos(angle) * startSpeed;
		speedY = yLaunchSpeed + (float) Math.sin(angle) * startSpeed;

		
		float launchSpeed = startSpeed;
		lifeTimeSeconds = range / launchSpeed + 5;
		
		startX = newX; startY = newY;
		
		this.x = newX + (float) Math.cos(angle) * length;
		this.y = newY + (float) Math.sin(angle) * length;
		backX = newX;
		backY = newY;
		oldBackX = backX;
		oldBackY = backY;
	}
	@Override
	public void doBulletMove(double seconds, double timeAtStartOfMoveSeconds) {
	
		super.doBulletMove(seconds, timeAtStartOfMoveSeconds);
		
		if(!dead && getClosestEnemy()!=null && Point2D.distance(x, y, startX, startY) > 40)
		{
			targetX = getClosestEnemy().getX();
			targetY = getClosestEnemy().getY();
			updateTurning(seconds);
			this.speedX = (float) (this.getSpeed() * Math.cos(angle));
			this.speedY = (float) (this.getSpeed() * Math.sin(angle));
		}

		//world.getParticleEmitters().add(new ParticleTrail(new Point2D.Float(x, y), CustomColor.PURPLE, 1, 2, false));
		
	}
	public float calcAngle(float xCoord, float yCoord) {
		float angle = 0;
		if (xCoord != 0) {
			angle = (float) Math.atan(yCoord / xCoord);
			if (xCoord < 0) {
				angle += (float) Math.PI;
			}
		} else if (yCoord > 0) {
			angle = (float) (Math.PI / 2);
		} else if (yCoord < 0) {
			angle = (float) (3 * Math.PI / 2);
		}
		if (angle < 0) {
			angle += (float) (2f * Math.PI);
		}
		assert angle >= 0 : "angle == " + angle + ", xCoord == " + xCoord + ", yCoord == " + yCoord;
		return angle;
	}
	public void updateTurning(double seconds) {
		
		float oldAngle = this.getAngle();
		if (!dead && getClosestEnemy()!=null) {
			//float targetGunAngle = (float) getAngle(mouseTargetX - getX(), mouseTargetY - getY());
			float targetAngle = calcAngle(getClosestEnemy().getX() - getX(), getClosestEnemy().getY() - getY());

			float angleToTurn = targetAngle - oldAngle;
			// Here we make sure angleToTurn is between -Math.PI and +Math.PI so 
			// that it's easy to know which way the gun should turn.
			// The maximum that angleToTurn could be now is + or - 2 * 2*Math.PI.
			if (angleToTurn < -Math.PI) {
				angleToTurn += (float) (2 * Math.PI);
				if (angleToTurn < -Math.PI) {
					angleToTurn += (float) (2 * Math.PI);
				}
			}
			if (angleToTurn > Math.PI) {
				angleToTurn -= (float) (2 * Math.PI);
				if (angleToTurn > Math.PI) {
					angleToTurn -= (float) (2 * Math.PI);
				}
			}
			assert angleToTurn >= -(float)Math.PI && angleToTurn <= (float)Math.PI : angleToTurn + ", " + Math.PI;
			float maxAngleChange = (float) (getMaxRotationSpeed() * seconds);
			if (angleToTurn > 0) {
				if (angleToTurn > maxAngleChange) {
					angle = oldAngle + maxAngleChange;
				} else {
					angle = oldAngle + angleToTurn;
				}
			} else {
				if (angleToTurn < -maxAngleChange) {
					angle = oldAngle - maxAngleChange;
				} else {
					angle = oldAngle + angleToTurn;
				}
			}
			if (getAngle() < 0) {
				angle += (float) (2 * Math.PI);
			}
			if (getAngle() >= 2 * Math.PI) {
				angle -= (float) (2 * Math.PI);
			}
			assert targetAngle >= 0 : targetAngle;
			assert getAngle() >= 0 : getAngle();
		}
	}
	public Player getClosestEnemy(){
		// modified from the bot method: for certain cases: ie. the 2nd closest enemy is shootable but 1st isn't:
		// this will shoot 2nd closest
		// bot method will not shoot at all.
		Player closestPlayer = null;
		double closestDist = Double.MAX_VALUE;
		ArrayList<Player> isShootable = new ArrayList<Player>();
		for(Player p : world.getPlayers())
		{
			if (p.isDead() || (p.isInvisible() && p.getGun().isFiring() == false) || p.getTeam() == this.getPlayer().getTeam()){
				continue;
			}
			if(isPlayerShootable(player))
				isShootable.add(p);
		}
		for (Player p : isShootable){
			
			if (Point2D.distance(p.getX(), p.getY(), getX(), getY()) < closestDist){
				closestDist = Point2D.distance(p.getX(), p.getY(), getX(), getY());
				closestPlayer = p;
			}
		}
		return closestPlayer;
	}
	protected boolean isPlayerShootable(Player p){
		ArrayList<Obstacle> obstacles = getWorld().getObstacles();
		for (int i = 0; i < obstacles.size(); i++) {
			Obstacle obstacle = obstacles.get(i);
			KPolygon shape = obstacle.getShape();
//			if (Point2D.Float.distance(x, y, shape.getCentre().x, shape.getCentre().y) > shape.getCircularBound() + radius) {
//				continue;
//			}
			Point2D.Float[] points = shape.getPoints();
			for (int j = 0; j < points.length; j++) {
				int jPlus = (j + 1 == points.length ? 0 : j + 1);
				if (Line2D.linesIntersect(getX(), getY(), p.getX(), p.getY(), points[j].x, points[j].y, points[jPlus].x, points[jPlus].y)) {
					return false;
				}
			}
		}
		return true;
	}

	//transient BasicStroke stroke = new BasicStroke(3);
	@Override
	public void render(ViewPane viewPane) {
		Graphics2D g = viewPane.getBackImageGraphics2D();
		
		g.setColor(this.player.getTeam().getColor());
		g.draw(new Ellipse2D.Float(getX() - radius * 1.2f, getY() - radius * 1.2f, radius * 2f * 1.2f, radius * 2f * 1.2f));
				
		g.setColor(CustomColor.PURPLE);		
		g.fill(new Ellipse2D.Float(getX() - radius* .8f, getY() - radius* .8f, radius * 2f* .8f, radius * 2f* .8f));
	}

	@Override
	public float getDamage(){
		return damage;
	}

	@Override
	public double getLifeTimeSeconds(){
		return lifeTimeSeconds;
	}
	@Override
	public float getLength(){
		return length;
	}
	@Override
	public float getCanNotHitOwnPlayerTimeSeconds(){
		return canNotHitOwnPlayerTimeSeconds;
	}
	public float getAngle() {
		return this.angle;
	}
	public float getMaxRotationSpeed() {
		return maxRotationSpeed;
	}
	
	
}


SydneyShooter/src/sydneyengine/ui/HelpPane.java
/*
 * HelpPane.java
 *
 * Created on 6 January 2008, 22:12
 */

package sydneyengine.ui;

import javax.swing.JInternalFrame;

import sydneyengine.Controller;
import sydneyengine.ServingController;
import sydneyengine.shooter.ViewPane;

/**
 *
 * @author  woodwardk
 */
public class HelpPane extends javax.swing.JPanel {
	ViewPane v;
	JInternalFrame f;
	/** Creates new form HelpPane */
	public HelpPane(final ViewPane v, final JInternalFrame f) {
		this.v = v;
		this.f = f;
		Controller c = v.getGameFrame().getController();
		initComponents();
		String serverIPAddressAndSocketExplanationString = "";
		if (c instanceof ServingController){
			serverIPAddressAndSocketExplanationString += "You are the server and your IP address and port is: "+((ServingController)c).getConnectionServer().getInetSocketAddress().toString();
			serverIPAddressAndSocketExplanationString += "\n";
		}
		String text = 
				serverIPAddressAndSocketExplanationString+
				"Aim: Capture the flag and bring it back to your team's nearest base to score a point.\n" +
				"\n======[ Controls ]======\n"+
				"Move with Arrow keys or W, A, S, D\n"+
				"Left mouse click: fires weapon\n"+
				"R: reloads weapon\n"+
				"Number keys (1-9) or Q/E or Mouse wheel: selects available weapons\n"+
				"`[funny key above TAB]: turn on/off player names and hit points\n"+
				"Enter or Shift+Enter: Activate chat, then and once you've typed the message, press enter again to send. "
				+ "\nIf shift is down when you first press enter, the 'send to allies' option is ticked\n"+
				"Page up or down: zoom in/ out\n"+
				"Escape: displays menu.\n"+
				
				"\n======[ Weapons ]======\n"+
				"\nPistol - Basic weapon with low damage.\n"+
				"\nMachine Gun - Standard rapid-firing rotating gatling.\n"+
				"\nFlamethrower - Close-range weapon that sprays napalm that will stick around for a while, useful for area denial. "
				+ "The napalm will cause damage to anyone that walks over it.\n"+
				"\nShotgun - Dual-shot weapon that fires a cloud of pellets. Most effective at close ranges\n"+
				"\nSniping Rifle - Accurate for long-distance one-hit kills, but slow-firing.\n"+
				"\nTranquilizer Gun - Fires in short bursts, stunning enemies and slowing them and making them unable to fire.\n"+
				"\nRocket Launcher - launches rockets that have a large explosion radius. More damage is dealt to objects that"
				+ "are closer to the epicenter of the explosion.\n"+
				"\nHoming Gun - launches three alien-technology orbs that will fire like a regular bullet unless there is an enemy"
				+ "in front of it, in which case it will automatically follow that target. The orbs are relatively slow, but are"
				+ "still useful around corners and against dug-in defenders.\n"+
				"\nNail Gun - An alternative to the machine gun, that spits out 3 nails at once very rapidly,"
				+ "but watch out for your ammo depletion, since it comes with a small clip considering its rate of fire.\n"+
				"\n***Note***\n"+
				"Both the Flamethrower and the Rocket Launcher are Variable-Range weapons, that is, their bullets can detonate or stick at a "
				+ "certain range away from the weapon by using the mouse to aim.\n"+
				
				"\n======[ Items ]======\n"+
				"\nHealthpack - Restores 33% health.\n"+
				"\nGodly Armor - Grants invincibility against all weapons for 15 seconds, but player can still be tranquilized.\n"+
				"\nSpeed Shoes - Significantly speeds up travel.\n"+
				"\nInvisibility Shroud - Enemies can't see and target you for 15 seconds.\n"+
				"\n***Note***\n"+
				"You can only use one of either the Godly Armor or the Invisibility Shroud, but not both. But, the Speedy Shoes can be"
				+ "use with anything.\n"+
				
				"\n======[ Extra ]======\n"+ 
				"\nThe wi-fi on laptops may cause noticeable lag over local LAN, but still very playable.\n" +
				
				"\n\nThis was slightly modified from the excellent Sydney Engine v. 0.2, made by Keith Woodward, "
				+ "whom can find me on the forums at www.javagaming.org or contact through keithphw@hotmail.com\n"+
				"If you'd like to expand on this game, you can download the source code of Sydney Engine on www.javagaming.org, "
				+ "just search for 'SydneyEngine'.";
		jTextPane1.setText(text);
	}
	
	/** This method is called from within the constructor to
	 * initialize the form.
	 * WARNING: Do NOT modify this code. The content of this method is
	 * always regenerated by the Form Editor.
	 */
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        jPanel1 = new javax.swing.JPanel();
        jLabel2 = new javax.swing.JLabel();
        returnToGameButton = new javax.swing.JButton();
        jScrollPane1 = new javax.swing.JScrollPane();
        jTextPane1 = new javax.swing.JTextPane();

        jLabel2.setFont(new java.awt.Font("Franklin Gothic Medium", 0, 36));
        jLabel2.setHorizontalAlignment(javax.swing.SwingConstants.CENTER);
        jLabel2.setText("Defiance"); // NOI18N

        returnToGameButton.setText("Return to game"); // NOI18N
        returnToGameButton.addActionListener(new java.awt.event.ActionListener() {
            @Override
			public void actionPerformed(java.awt.event.ActionEvent evt) {
                returnToGameButtonreturnToGame(evt);
            }
        });

        jTextPane1.setEditable(false);
        jScrollPane1.setViewportView(jTextPane1);

        org.jdesktop.layout.GroupLayout jPanel1Layout = new org.jdesktop.layout.GroupLayout(jPanel1);
        jPanel1.setLayout(jPanel1Layout);
        jPanel1Layout.setHorizontalGroup(
            jPanel1Layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(jPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .add(jPanel1Layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                    .add(jScrollPane1, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, 438, Short.MAX_VALUE)
                    .add(org.jdesktop.layout.GroupLayout.TRAILING, jLabel2, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, 438, Short.MAX_VALUE)
                    .add(org.jdesktop.layout.GroupLayout.TRAILING, returnToGameButton, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, 438, Short.MAX_VALUE))
                .addContainerGap())
        );
        jPanel1Layout.setVerticalGroup(
            jPanel1Layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(jPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .add(jLabel2)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.RELATED)
                .add(jScrollPane1, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, 330, Short.MAX_VALUE)
                .addPreferredGap(org.jdesktop.layout.LayoutStyle.UNRELATED)
                .add(returnToGameButton, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE, 50, org.jdesktop.layout.GroupLayout.PREFERRED_SIZE)
                .addContainerGap())
        );

        org.jdesktop.layout.GroupLayout layout = new org.jdesktop.layout.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(0, 458, Short.MAX_VALUE)
            .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                .add(layout.createSequentialGroup()
                    .add(0, 0, 0)
                    .add(jPanel1, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .add(0, 0, 0)))
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
            .add(0, 460, Short.MAX_VALUE)
            .add(layout.createParallelGroup(org.jdesktop.layout.GroupLayout.LEADING)
                .add(layout.createSequentialGroup()
                    .add(0, 0, 0)
                    .add(jPanel1, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, org.jdesktop.layout.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .add(0, 0, 0)))
        );
    }// </editor-fold>//GEN-END:initComponents

	private void returnToGameButtonreturnToGame(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_returnToGameButtonreturnToGame
		v.getGameFrame().getDesktopPane().remove(f);
		f.dispose();
		v.requestFocus();
	}//GEN-LAST:event_returnToGameButtonreturnToGame
	
	
    // Variables declaration - do not modify//GEN-BEGIN:variables
    protected javax.swing.JLabel jLabel2;
    protected javax.swing.JPanel jPanel1;
    protected javax.swing.JScrollPane jScrollPane1;
    protected javax.swing.JTextPane jTextPane1;
    protected javax.swing.JButton returnToGameButton;
    // End of variables declaration//GEN-END:variables
	
}
